### PR TITLE
Async events improve

### PR DESCRIPTION
**Pull request by ramon-bernardo**
### Melhorias
- Prefira [crossbeam-channel](https://github.com/crossbeam-rs/crossbeam/tree/master/crossbeam-channel) ao invés de [std mpsc](https://doc.rust-lang.org/std/sync/mpsc/index.html).
- Renomeia `EventSender` para `AsyncEventSender` e `EventReceiver` para `AsyncEventReceiver`.
- Adiciona `AsyncEventSet`, possibilitando ordenar os sistemas.
- Renomeia `channel_to_event `para `unbounded_channel_to_event`.
- Prefira executar o mapeamento entre mensagem do canal para evento no Scheduler `PreUpdate` ao invés de `Update`.

Meu analyzer ta bugado com os 600 pacotes do projeto, além dos arquivos proto, pode ser que precise de algum polimento no código, além da doc do _SystemSet_.

### CODE REVIEW

File: ryot/Cargo.toml [hunk 1]
```diff
@@ -16,6 +16,7 @@ readme = "../README.md"
 async-std = "1.12.0"
 derive_more = "0.99.17"
 serde_repr = "0.1"
+crossbeam-channel = "0.5.12"
```
Comment (ramon-bernardo) [inline ryot/Cargo.toml:19]: Se preferir, posso mudar para o workspace.

**Comment by lgrossi** [comment]
Hi, thank you for collaborating! Can you summarize the main differences between crossbeam and mpsc?
It's the first time I'm seeing crossbeam, so I'm curious about the trade offs.

**Comment by ramon-bernardo** [comment]
Até um tempo atrás, crossbeam-channel e std::sync::mpsc tinham uma grande diferença de desempenho. Pesquisando um pouco mais, parece que foi [mesclado](https://github.com/rust-lang/rust/pull/93563) o código do crossbeam no mpsc std, no entanto, as melhorias e correções sempre chegam mais cedo em crossbeam-channel do que em std. Por exemplo, [esta correção](https://github.com/rust-lang/rust/pull/121646) ainda não está em rust 1.78.0, e já foi corrigido no crossbeam.

O uso de crossbeam-channel também nos da a possibilidade de usar: [try_send](https://docs.rs/crossbeam-channel/0.5.12/crossbeam_channel/struct.Sender.html#method.try_send), [send_timeout](https://docs.rs/crossbeam-channel/0.5.12/crossbeam_channel/struct.Sender.html#method.send_timeout), [send_deadline](https://docs.rs/crossbeam-channel/0.5.12/crossbeam_channel/struct.Sender.html#method.send_deadline), [is_empty](https://docs.rs/crossbeam-channel/0.5.12/crossbeam_channel/struct.Sender.html#method.is_empty), [is_full](https://docs.rs/crossbeam-channel/0.5.12/crossbeam_channel/struct.Sender.html#method.is_full), [len](https://docs.rs/crossbeam-channel/0.5.12/crossbeam_channel/struct.Sender.html#method.len), [capacity](https://docs.rs/crossbeam-channel/0.5.12/crossbeam_channel/struct.Sender.html#method.capacity), [same_channel](https://docs.rs/crossbeam-channel/0.5.12/crossbeam_channel/struct.Sender.html#method.same_channel).

**Comment by luan** [comment]
Hi @ramon-bernardo,

Thanks for the effort you put in here and for presenting us with the crossbeam option. At this time, I don't see a compelling reason to add an external dependency when the standard library is doing a good enough job and we're not seeing any performance issues. We also don't need multi-consumer so that's another reason why sticking with mpsc is the current choice.

Another piece of feedback: please try to stick to English in public communications here on Github. This project is open source because we want to make it as inclusive as possible, and English is the language that enables the most amount of people to understand us here.

Once again, thank you for putting this work in, but at the time we're not going to take it.

**Comment by ramon-bernardo** [comment]
@luan, os benchmarks atuais do crossbeam-channel estão desatualizados, procurando um [mais recente](https://github.com/crossbeam-rs/crossbeam/pull/1080), percebe-se uma melhoria de até 40% em SPSC e 80% em MPSC.
image

**Comment by lgrossi** [comment]
> @luan, os benchmarks atuais do crossbeam-channel estão desatualizados, procurando um [mais recente](https://github.com/crossbeam-rs/crossbeam/pull/1080), percebe-se uma melhoria de até 40% em SPSC e 80% em MPSC.
> image
> 

As explained above, please stick to english. We will no longer reply messages that are not in english. Thank you again for your time and for contributing.
